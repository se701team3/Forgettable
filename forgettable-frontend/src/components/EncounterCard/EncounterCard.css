.MuiAvatarGroup-avatar {
    display: flex;
    align-items: center;
    max-height: 50px;
    max-width: 170px;
}