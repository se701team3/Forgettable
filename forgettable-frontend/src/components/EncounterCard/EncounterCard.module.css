@import '../../colours.css';

.Card, .Card_special{
    width: fit-content;
    height: fit-content;
    background-color: var(--card);
    box-shadow: none;
    border-radius: 13px;
    z-index: 1;
    user-select: none;
    color: var(--txt1);
}

.Card_special{
    border: 4px solid var(--prmry);
}

.Card:hover, .Card_special:hover{
    background-color: var(--hilit);
    cursor: pointer;
}

.Card_content{
    display: flex;
    flex-direction: column;
    height: 250px;
    width: 715px;
    justify-content: center;
    padding: 15px 45px 15px 45px;
}

.Initial_label_container{
    width: inherit;
    display: flex;
    align-items: flex-start;
    text-align: center;
    margin-bottom: 10px;
}

.Initial_label{
    background-color: var(--prmry);
    padding: 2px 8px 2px 8px;
    border-radius: 3px;
    font-size: var(--text-xsmall);
}

.Header{
    width: inherit;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

.Title{
    font-size: var(--text-card-title);
    font-weight: var(--font-semibold);
    max-height: 50px;
    max-width: 480px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}

.Profile_container{
    font-size: var(--text-normal);
    font-weight: var(--font-regular);
}

.Body{
    margin-top: 15px;
    height: 80px;
    text-align: left;
    overflow: hidden;
    text-overflow: ellipsis;
    font-size: var(--text-small);
    font-weight: var(--font-regular);
    color: var(--txt2);
}

.Footer{
    display: flex;
    flex-direction: row;
    gap: 40px;
    align-items: baseline;
    justify-content: space-between;
    width: inherit;
    margin-top: 15px;
}

.Date, .Location{
    display: flex;
    flex-direction: row;
    gap: 10px;
    align-items: flex-start;
    font-size: var(--text-small);
    font-weight: var(--font-regular);
    color: var(--txt2);
}

.Avatar{
    font-family: 'Poppins', sans-serif;
    font-weight: var(--font-regular);
    font-size: var(--text-normal);
}

.Avatar_inline{
    display: flex;
    flex-direction: row;
    align-items: center;
    cursor: auto;
}

.Avatar_multiple{
    width: fit-content;
}

.Avatar_name{
    margin-left: 10px;
}