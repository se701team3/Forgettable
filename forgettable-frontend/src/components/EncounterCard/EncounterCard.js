/* eslint-disable max-len */
import * as React from 'react';
import Card from '@mui/material/Card';
import CustomButton from '../CustomButton/CustomButton';
import {AvatarGroup, Avatar} from '@mui/material';
import classes from './EncounterCard.module.css';
import {stringAvatar} from '../../functions/helper';
import './EncounterCard.css';
import PropTypes from 'prop-types';
import {getDateString} from '../../functions/dateFormatter';

const DELETE = 'Delete';
const DATE_WE_MET = 'Date we met: ';
const WHERE_WE_MET = 'Where we met: ';

/*
 * Component for displaying information of encounter details.
 * This is the full version of the encounter card.
 *
 * Author: Raina Song (rainasong)
 */
const EncounterCard = (props) => {
  const {isInitialEncounter, title, persons, description, date, location, onClick, onDelete} = props;
  const isMultiplePerson = persons.length > 1;

  return (
    <div className={isInitialEncounter ? classes.Card_special : classes.Card}>
      <div className={classes.Card_content}>
        <div onClick={onClick}>
          {isInitialEncounter && <section className={classes.Initial_label_container}>
            <div className={classes.Initial_label}>First Encounter!</div>
          </section>}
          <section className={classes.Header}>
            <div className={classes.Title}>
              {title}
            </div>
            <div className={classes.Profile_container}>
              <AvatarGroup max={4}
                className={isMultiplePerson ? classes.Avatar_multiple : classes.Avatar_inline}
                data-testid="persons-avatar-group"
              >
                {persons.map((person, index) => {
                  return (
                    <div key={`${index}-container`}>
                      <Avatar key={index} alt={`${person.first_name} ${person.last_name}`} className={classes.Avatar} {...!person.img && stringAvatar(`${person.first_name} ${person.last_name}`)} src={person.img} />
                      {!isMultiplePerson && <div className={classes.Avatar_name}>
                        {person.first_name}
                      </div>}
                    </div>
                  );
                },
                )}
              </AvatarGroup>
            </div>
          </section>
          <section className={classes.Body}>
            {description}
          </section>
        </div>
        <section className={classes.Footer}>
          <div className={classes.Date}>
            {DATE_WE_MET}<div>{getDateString(date)}</div>
          </div>
          <div className={classes.Location}>
            {WHERE_WE_MET}<div>{location}</div>
          </div>
          <CustomButton btnText={DELETE} onClick={onDelete}/>
        </section>
      </div>
    </div>
  );
};

EncounterCard.propTypes = {
  id: PropTypes.string,
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  persons: PropTypes.arrayOf(PropTypes.shape({
    first_name: PropTypes.string.isRequired,
    last_name: PropTypes.string.isRequired,
    img: PropTypes.string,
  })),
  date: PropTypes.instanceOf(Date),
  location: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  isInitialEncounter: PropTypes.bool.isRequired,
};

export default EncounterCard;
