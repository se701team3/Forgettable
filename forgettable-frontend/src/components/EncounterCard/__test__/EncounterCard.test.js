import TestRenderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import EncounterCard from '../EncounterCard';
import {render, screen} from '@testing-library/react';

const persons = [
  {name: 'Kent Dodds'},
  {name: 'Jed Watson'},
  {name: 'Marley George'},
  {name: 'Jack Winston'},
  {name: 'John Doe'},
];
const title = 'Example Title';
const description = 'Example Description';
const location = 'Example Location';
const isInitialEncounter = false;

it('renders EncounterCard UI to test correct hierarchy', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<EncounterCard
    title={title}
    description={description}
    location={location}
    persons={persons}
    onClick={jest.fn()}
    onDelete={jest.fn()}
    isInitialEncounter={isInitialEncounter}
  />);
  const result = renderer.getRenderOutput();

  expect(result).toMatchSnapshot();
});

it('Title displays correctly', () => {
  render(<EncounterCard
    title={title}
    description={description}
    location={location}
    persons={persons}
    onClick={jest.fn()}
    onDelete={jest.fn()}
    isInitialEncounter={isInitialEncounter}
  />);
  const titleElement = screen.getByText(/example title/i);

  expect(titleElement).toBeInTheDocument();
});

it('Description displays correctly', () => {
  render(<EncounterCard
    title={title}
    description={description}
    location={location}
    persons={persons}
    onClick={jest.fn()}
    onDelete={jest.fn()}
    isInitialEncounter={isInitialEncounter}
  />);
  const descriptionElement = screen.getByText(/example description/i);

  expect(descriptionElement).toBeInTheDocument();
});

it('Location displays correctly', () => {
  render(<EncounterCard
    title={title}
    description={description}
    location={location}
    persons={persons}
    onClick={jest.fn()}
    onDelete={jest.fn()}
    isInitialEncounter={isInitialEncounter}
  />);
  const locationElement = screen.getByText(/example description/i);

  expect(locationElement).toBeInTheDocument();
});

it('Renders the correct number of avatars when persons number is greater than max (4)', () => {
  const maxAvatarNum = 4;
  render(<EncounterCard
    title={title}
    description={description}
    location={location}
    persons={persons}
    onClick={jest.fn()}
    onDelete={jest.fn()}
    isInitialEncounter={isInitialEncounter}
  />);
  const avatarGroupElement = screen.getByTestId('persons-avatar-group');

  expect(avatarGroupElement.childElementCount).toBe(maxAvatarNum);
});

it('Renders the correct number of avatars when persons number is lower than max (4) and not one', () => {
  const persons = [
    {name: 'Kent Dodds'},
    {name: 'Jed Watson'},
    {name: 'Marley George'},
  ];
  render(<EncounterCard
    title={title}
    description={description}
    location={location}
    persons={persons}
    onClick={jest.fn()}
    onDelete={jest.fn()}
    isInitialEncounter={isInitialEncounter}
  />);
  const avatarGroupElement = screen.getByTestId('persons-avatar-group');

  expect(avatarGroupElement.childElementCount).toBe(persons.length);
});

it('Renders the correct number of avatars when persons number is one', () => {
  const persons = [
    {name: 'Kent Dobbs'},
  ];
  render(<EncounterCard
    title={title}
    description={description}
    location={location}
    persons={persons}
    onClick={jest.fn()}
    onDelete={jest.fn()}
    isInitialEncounter={isInitialEncounter}
  />);
  const avatarGroupElement = screen.getByTestId('persons-avatar-group');

  expect(avatarGroupElement.childElementCount).toBe(persons.length);
});

it('Date is displayed in the correct format', () => {
  const date = new Date(1646965503063);

  render(<EncounterCard
    title={title}
    description={description}
    location={location}
    persons={persons}
    onClick={jest.fn()}
    onDelete={jest.fn()}
    date={date}
    isInitialEncounter={isInitialEncounter}
  />);

  const dateElement = screen.getByText('11/03/2022');

  expect(dateElement).toBeInTheDocument();
});

it('Displays date placeholder correctly when there is no date specified', () => {
  render(<EncounterCard
    title={title}
    description={description}
    location={location}
    persons={persons}
    onClick={jest.fn()}
    onDelete={jest.fn()}
    isInitialEncounter={isInitialEncounter}
  />);
  const dateElement = screen.getByText(/unknown/i);

  expect(dateElement).toBeInTheDocument();
});

it('Displays the first encounter label when it is the first encounter', () => {
  const isInitialEncounter = true;

  render(<EncounterCard
    title={title}
    description={description}
    location={location}
    persons={persons}
    onClick={jest.fn()}
    onDelete={jest.fn()}
    isInitialEncounter={isInitialEncounter}
  />);

  const firstEncounterLabelElement = screen.getByText(/first encounter!/i);

  expect(firstEncounterLabelElement).toBeInTheDocument();
});
