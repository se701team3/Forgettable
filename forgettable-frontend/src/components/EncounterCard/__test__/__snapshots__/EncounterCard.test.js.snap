// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders EncounterCard UI to test correct hierarchy 1`] = `
<div
  className="Card"
  onClick={[Function]}
>
  <div
    className="Card_content"
  >
    <section
      className="Header"
    >
      <div
        className="Title"
      >
        Example Title
      </div>
      <div
        className="Profile_container"
      >
        <CustomAvatar
          persons={
            Array [
              Object {
                "name": "Kent Dodds",
              },
              Object {
                "name": "Jed Watson",
              },
              Object {
                "name": "Marley George",
              },
              Object {
                "name": "Jack Winston",
              },
              Object {
                "name": "John Doe",
              },
            ]
          }
        />
      </div>
    </section>
    <section
      className="Body"
    >
      Example Description
    </section>
    <section
      className="Footer"
    >
      <div
        className="Date"
      >
        Date we met: 
        <div>
          <UnknownDetail />
        </div>
      </div>
      <div
        className="Location"
      >
        Where we met: 
        <span>
          Example Location
        </span>
      </div>
      <CustomButton
        btnText="Delete"
        onClick={[Function]}
      />
    </section>
  </div>
</div>
`;
